app/frontend/src/components/AvatarInput.tsx

```diff
@@ -1,54 +1,166 @@
+  SELECT_AN_IMAGE,
+  UPLOAD_PENDING_ERROR,
+} from "components/constants";
+import IconButton from "components/IconButton";
+import { CheckIcon, CrossIcon } from "components/Icons";
+import React, { useRef, useState } from "react";
+import { Control, useController } from "react-hook-form";
+import { useMutation } from "react-query";
+import { service } from "service";
+import { ImageInputValues } from "service/api";
 
-import { ImageInput, ImageInputProps, ImageInputValues } from "./ImageInput";
-
-const useStyles = makeStyles(() => ({
+const useStyles = makeStyles((theme) => ({
+  root: {
+    display: "flex",
+  },
   avatar: {
-    "& img": { objectFit: "contain" },
+    "& img": { objectFit: "cover" },
+  },
+  confirmationButtonContainer: {
+    display: "flex",
+    flexDirection: "column",
+    justifyContent: "center",
+    "& > * + *": {
+      marginTop: theme.spacing(1),
+    },
+  },
+  input: {
+    display: "none",
+  },
+  label: {
+    alignItems: "center",
+    display: "flex",
+    justifyContent: "center",
+    width: "fit-content",
+  },
+  loading: {
+    position: "absolute",
   },
 }));
 
-export interface AvatarInputProps
-  extends Omit<ImageInputProps, "children" | "onChange"> {
-  src?: string;
-  onChange: (key: string) => void;
+interface AvatarInputProps {
+  control: Control;
+  id: string;
   className?: string;
+  name: string;
+  initialPreviewSrc?: string;
   userName?: string;
 }
 
 export function AvatarInput({
+  control,
   id,
   name,
-  src,
+  initialPreviewSrc,
   userName,
   className,
-  onChange,
-  ...rest
 }: AvatarInputProps) {
   const classes = useStyles();
-  const [imageUrl, setImageUrl] = useState(src);
-  const handleChange = (values: ImageInputValues) => {
-    const randomInt = Math.floor(Math.random() * 100); // force reload onChange
-    setImageUrl(values.thumbnail_url + "?rand=" + randomInt);
-    onChange && onChange(values.key);
+  //this ref handles the case where the user uploads an image, selects another image,
+  //but then cancels - it should go to the previous image rather than the original
+  const confirmedUpload = useRef<ImageInputValues>();
+  const [imageUrl, setImageUrl] = useState(initialPreviewSrc);
+  const [file, setFile] = useState<File | null>(null);
+  const mutation = useMutation<ImageInputValues, Error>(() =>
+    file ? service.api.uploadFile(file) : Promise.reject(INVALID_FILE)
+  );
+  const isConfirming = !mutation.isLoading && file !== null;
+  const { field } = useController({
+    name,
+    control,
+    defaultValue: "",
+    rules: {
+      validate: () => !isConfirming || UPLOAD_PENDING_ERROR,
+    },
+  });
+
+  const handleChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (!event.target.files?.length) return;
+    const file = event.target.files[0];
+    const base64 = await new Promise<string>((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onload = () => resolve(reader!.result as string);
```

Comment:
```suggestion
      reader.onload = () => resolve(reader.result as string);
```

Reply:
Typescript doesn't know it can't be undefined, since it starts as undefined and the event can be triggered from anything in theory

Reply:
`reader` won't be undefined though as you instantiated on the line above? I checked locally and it didn't give me an error without it.

Reply:
Oh quite right